### PR TITLE
[SYCL] Optimize vec<bfloat> math builtins

### PR DESCRIPTION
**Pull request by uditagarwal97**
Followup and blocked by: https://github.com/intel/llvm/pull/14105

Currently, `vec<bfloat>` math builtins do element-by-element operations.
This PR optimize `vec<bfloat>` math builtins by:
(1) Converting `vec<bfloat>` to `vec<float>`.
(2) Do the operation on `vec<float>` (which uses Spirv built-ins underneath for optimized vector operations).
(3) Convert back the return value to `vec<bfloat>`.

Look at the beautiful diff in `check_device_code/vector/vector_bf16_builtins.cpp` to visualize the device code generated before and after this optimization.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:67]:
```suggestion
  // Convert BFloat16 vector to float vec and call isnan().
```

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:70]: This should be `Res` but I see the rest of the file using `res`, so honestly I'm not sure what's the best way to proceed here. @intel/llvm-reviewers-runtime any opinion?

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:411]: https://www.llvm.org/docs/CodingStandards.html#comment-formatting:

> `//` is preferred for normal comments.

File: sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp
Comment (maarquitos14) [inline sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp:34]: https://www.llvm.org/docs/CodingStandards.html#comment-formatting:

> `//` is preferred for normal comments.

File: sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp
Comment (maarquitos14) [inline sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp:38]: https://www.llvm.org/docs/CodingStandards.html#comment-formatting:

> `//` is preferred for normal comments.

File: sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp
Comment (maarquitos14) [inline sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp:58]: https://www.llvm.org/docs/CodingStandards.html#comment-formatting:

> `//` is preferred for normal comments.

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:70]: Applies to other occurrences as well.

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (uditagarwal97) [reply sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:411]: AFAIK, we can not use `//` inside macro. https://stackoverflow.com/questions/24751387/can-i-comment-multi-line-macros

Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:411]: Oh, totally right. Didn't realize it was within the macro. You can ignore this.

File: sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp
Comment (uditagarwal97) [reply sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp:70]: > This should be `Res` but I see the rest of the file using `res`, so honestly I'm not sure what's the best way to proceed here. @intel/llvm-reviewers-runtime any opinion?

There's an inconsistency in the variable naming convention used in this file. Since LLVM suggests using CamelCase for variable names (https://www.llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly), I could use `Res` instead.

Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp]: I'd like it better if you did, at least let's not introduce any more errors. Thanks!

**Review by maarquitos14 — APPROVED** [review]
LGTM.

**Comment by uditagarwal97** [comment]
The failure in Gen12 Windows is unrelated and likely an infrastructure issue. 

@intel/llvm-gatekeepers the PR is ready to be merged!

**Comment by uditagarwal97** [comment]
> The failure in Gen12 Windows is unrelated and likely an infrastructure issue.
> 
> @intel/llvm-gatekeepers the PR is ready to be merged!

Ahh.. I overlooked the pending review from @intel/dpcpp-tools-reviewers

**Review by sarnex — APPROVED** [review]
sycl/test-e2e/BFloat16/bfloat16_vec_builtins.cpp LGTM, seems like whitespace only changes

**Comment by sarnex** [comment]
Gen12 issue is https://github.com/intel/llvm/issues/14248 but I restarted the job, I would prefer we run the tests before merging so we don't break postcommit.

File: sycl/test/check_device_code/vector/vector_bf16_builtins.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/vector/vector_bf16_builtins.cpp:39]: Why new is better? Do you have any performance measurements? I see we're doing extra converts and also I'm not sure this "optimization" doesn't change the resulting value vs native bfloat16 compute.

Comment (uditagarwal97) [reply sycl/test/check_device_code/vector/vector_bf16_builtins.cpp:39]: > Why new is better?

Previously, on device, we were doing element-by-element computation in math builitins (for `vec<bfloat16>`). Now, after this change, we convert BF16 `sycl::vec` to `sycl::vec<float>`, perform computations on `sycl::vec<float>` using spirv builitins (which should vectorize operations underneath), and convert the result back to `vec<bfloat16>`. On device, the conversion between `vec<float>` and `vec<bfloat16>` happens using OpConvertBF16ToFINTEL and OpConvertFToBF16INTEL which will also result in vectorized conversions.
Note that, currently, for BF16 scalar support in math builtins, we anyway convert back and forth to float: https://github.com/uditagarwal97/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp#L342

> Do you have any performance measurements?

I don't think it's necessary - generated device code is clear enough.

> I'm not sure this "optimization" doesn't change the resulting value vs native bfloat16 compute

I'm not exactly sure what you meant by this. Looking at the device code, the return value of `__spirv_ocl_fmin` is converted to `vec<bfloat16>` and returned, which seems correct to me.

Comment (aelovikov-intel) [reply sycl/test/check_device_code/vector/vector_bf16_builtins.cpp:39]: We vectorize over entire sub-group, not across vector elements. I doubt the improvement in the operation itself would be as significant as you expect it to be and possibly not big enough to justify double convert (bf16->float->bf16).

Comment (uditagarwal97) [reply sycl/test/check_device_code/vector/vector_bf16_builtins.cpp:39]: Regarding overhead of convert, we do the conversion from `float` <--> `bfloat` even before this optimization. In fact, earlier we used to do element-by-element conversion but now we use OpConvertBF16ToFINTEL etc. for vectorized conversions, which IIUC, should strictly be faster.

**Pull request merged**
